### PR TITLE
Story/quad 46

### PR DESCRIPTION
**Pull request by suprajkolluri**
resolved merge conflicts and refactored the code based on review comments

### CODE REVIEW

File: Quadriga/src/main/java/edu/asu/spring/quadriga/aspects/DictionaryAuthorization.java [hunk 1]
```diff
@@ -53,6 +49,9 @@ public boolean chkAuthorization(String userName, String accessObjectId,
         IDictionary dictionary = dictonaryManager
                 .getDictionaryDetails(accessObjectId);
 
+        if(dictionary == null){
+        	return false;
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/aspects/DictionaryAuthorization.java:53]: I think you need to throw an exception here like for workspaces. Or am I misreading this? If someone tries to access a dictionary that doesn't exist than that is an error not authorization denied.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/aspects/RestAccessAspect.java [hunk 2]
```diff
@@ -11,92 +11,91 @@
...
+
+                } else {
+                    haveAccess = authorization.chkAuthorizationByRole(userName, policy.userRole());
+                }
+
+                if (haveAccess) {
+                    break;
+                }
+            }
+        } catch (QuadrigaAccessException e) {
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/aspects/RestAccessAspect.java:90]: this catch block should be where the exception is actually thrown (line 80?), not for the whole block.

Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/aspects/RestAccessAspect.java:90]: also this should be a not found exception, not forbidden, shouldn't it? the exception is only thrown if an object for the provided id doesn't exist, right?

File: Quadriga/src/main/java/edu/asu/spring/quadriga/exceptions/RestException.java [hunk 3]
```diff
@@ -40,5 +39,10 @@ public RestException(int errorno, String message) {
 		
 	}
 
+    public RestException(int errorno, Throwable error) {
+        super(error);
+        errorcode = errorno;
+    }
+
 	
 }
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/exceptions/RestException.java:48]: format the class

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 4]
```diff
@@ -73,425 +75,391 @@
...
+        } catch (ResourceNotFoundException e) {
+            throw new RestException(404, e);
+        } catch (ParseErrorException e) {
+            throw new RestException(400, e);
+        } catch (MethodInvocationException e) {
+            throw new RestException(400, e);
+        } catch (QuadrigaStorageException e) {
+            throw new RestException(500, e);
+        } catch (Exception e) {
+            throw new RestException(400, e);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:156]: do we really need this? what exception are you catching here? It's very bad practice to have a catch-it-all clause.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 5]
```diff
@@ -73,425 +75,391 @@
...
+            VelocityContext context = new VelocityContext(restVelocityFactory.getVelocityContext());
+            context.put("list", collectionsList);
+
+            StringWriter writer = new StringWriter();
+            template.merge(context, writer);
+            return new ResponseEntity<String>(writer.toString(), HttpStatus.OK);
+        } catch (ResourceNotFoundException e) {
+            throw new RestException(404, e);
+        } catch (ParseErrorException e) {
+            throw new RestException(400, e);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:150]: where is that thrown? by the velocity engine? If so, I don't think that should be a 400, which would indicate the client sent a bad request, but a 500, given that something would be wrong the template and there is nothing the client could do about it.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 6]
```diff
@@ -73,425 +75,391 @@
...
+        } catch (MethodInvocationException e) {
+            throw new RestException(400, e);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:152]: similar to the above comment. where is this thrown?

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 7]
```diff
@@ -73,425 +75,391 @@
...
+     * @throws JAXBException
+     * @throws QuadrigaAccessException
+     * @throws QuadrigaStorageException
+     * @throws RestException
+     */
+    @RestAccessPolicies({
+            @ElementAccessPolicy(type = CheckedElementType.CONCEPTCOLLECTION_REST, paramIndex = 1, userRole = {
+                    RoleNames.ROLE_CC_COLLABORATOR_ADMIN, RoleNames.ROLE_CC_COLLABORATOR_READ_WRITE }) })
+    @RequestMapping(value = "rest/syncconcepts/{conceptCollectionID}", method = RequestMethod.POST)
+    public ResponseEntity<String> addConceptsToConceptColleciton(
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:192]: please correct spelling mistake in method name

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 8]
```diff
@@ -73,425 +75,391 @@
...
+            throw new RestException(404, e);
+        } catch (ParseErrorException e) {
+            throw new RestException(400, e);
+        } catch (MethodInvocationException e) {
+            throw new RestException(400, e);
+        } catch (QuadrigaStorageException e) {
+            throw new RestException(500, e);
+        } catch (Exception e) {
+            throw new RestException(400, e);
+        }
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:296]: see above comments

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 9]
```diff
@@ -73,425 +75,391 @@
...
+            Unmarshaller unmarshaller = context.createUnmarshaller();
+            unmarshaller.setEventHandler(new javax.xml.bind.helpers.DefaultValidationEventHandler());
+            InputStream is = new ByteArrayInputStream(xml.getBytes());
+            response1 = unmarshaller.unmarshal(new StreamSource(is), QuadrigaConceptReply.class);
+        } catch (JAXBException e) {
+            logger.error("Error in unmarshalling", e);
+            String errorMsg = restMessage.getErrorMsg("Error in unmarshalling", request);
+            return new ResponseEntity<String>(errorMsg, HttpStatus.BAD_REQUEST);
+        }
+        if (response1 == null) {
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:368]: according to the documentation, unmarshal never returns null. you can remove the if condition.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java [hunk 10]
```diff
@@ -73,425 +75,391 @@
...
+            throw new RestException(400, e);
+        } catch (MethodInvocationException e) {
+            throw new RestException(400, e);
+        } catch (QuadrigaStorageException e) {
+            throw new RestException(500, e);
+        } catch (QuadrigaAccessException e) {
+            throw new RestException(403, e);
+        } catch (Exception e) {
+            throw new RestException(400, e);
+        }
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:461]: see above comments

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/DictionaryRestController.java [hunk 11]
```diff
@@ -74,391 +76,342 @@
...
-
-			}
-			
-			response.setStatus(200);
-			response.setContentType(accept);
-			return "Success";
-		}
-	}
-	
+    private static final Logger logger = LoggerFactory.getLogger(DictionaryRestController.class);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/DictionaryRestController.java:79]: similar to the other class, please check the returned error messages

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/DictionaryRestController.java [hunk 12]
```diff
@@ -74,391 +76,342 @@
...
+            response1 = unmarshaller.unmarshal(new StreamSource(is), QuadrigaDictDetailsReply.class);
+        } catch (JAXBException e) {
+            logger.error("Error in unmarshalling", e);
+            String errorMsg = errorMessageRest.getErrorMsg("Error in unmarshalling", request);
+            return new ResponseEntity<String>(errorMsg, HttpStatus.BAD_REQUEST);
+        }
+        if (response1 == null) {
+            String errorMsg = errorMessageRest.getErrorMsg("Dictionary XML is not valid", request);
+            return new ResponseEntity<String>(errorMsg, HttpStatus.BAD_REQUEST);
+        }
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/DictionaryRestController.java:293]: response1 should never be null

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/DictionaryRestController.java [hunk 13]
```diff
@@ -74,391 +76,342 @@
...
+            Unmarshaller unmarshaller = context.createUnmarshaller();
+            unmarshaller.setEventHandler(new javax.xml.bind.helpers.DefaultValidationEventHandler());
+            InputStream is = new ByteArrayInputStream(xml.getBytes());
+            response1 = unmarshaller.unmarshal(new StreamSource(is), QuadrigaDictDetailsReply.class);
+        } catch (JAXBException e) {
+            logger.error("Error in unmarshalling", e);
+            String errorMsg = errorMessageRest.getErrorMsg("Error in unmarshalling", request);
+            return new ResponseEntity<String>(errorMsg, HttpStatus.BAD_REQUEST);
+        }
+        if (response1 == null) {
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/DictionaryRestController.java:381]: same here

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/DspaceRestController.java [hunk 14]
```diff
@@ -68,383 +67,424 @@
...
-
-			return dspaceProperties.getProperty("dspace.dspace_url")+dspaceProperties.getProperty("dspace.bitstream_url")+fileid+dspaceProperties.getProperty("dspace.xml")+dspaceProperties.getProperty("dspace.?")+
-					dspaceProperties.getProperty("dspace.api_key")+dspacekey.getPublicKey()+dspaceProperties.getProperty("dspace.&")+
-					dspaceProperties.getProperty("dspace.api_digest")+digestKey;
-		}
-
-		//No authentication provided
-		return dspaceProperties.getProperty("dspace.dspace_url")+dspaceProperties.getProperty("dspace.bitstream_url")+fileid+dspaceProperties.getProperty("dspace.xml");
-	}
+    private static final Logger logger = LoggerFactory.getLogger(DspaceRestController.class);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/DspaceRestController.java]: same for this class, please check error codes

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java [hunk 15]
```diff
@@ -48,661 +48,572 @@
...
+
+    @Autowired
+    private IEditingNetworkAnnotationManager editingNetworkAnnoManager;
+
+    @Autowired
+    private IRestVelocityFactory restVelocityFactory;
+
+    @Autowired
+    private IUserManager userManager;
+
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/NetworkRestController.java:80]: check error messages

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/WorkspaceRestController.java [hunk 16]
```diff
@@ -58,7 +60,7 @@
 @Controller
 public class WorkspaceRestController {
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/WorkspaceRestController.java:61]: please check error messages

File: Quadriga/src/main/resources/locale/errormessages.properties [hunk 17]
```diff
@@ -1,6 +1,6 @@
+error_message_400 = The server cannot or will not process the request due to a client error.
 error_message_401 = Unauthorized Access. Please check login credentials.
 error_message_403 = Not authorized to access this page.
-error_message_404 = Resource not found
-error_message_405 = Database is down!! People are working on it
-error_message_406 = Unknown error occurred please contact the administrator
-
+error_message_404 = Resource not found.
+error_message_406 = Unknown error occurred please contact the administrator.
```
Comment (jdamerow) [inline Quadriga/src/main/resources/locale/errormessages.properties:5]: 406 is not an unknown error, is it?

File: Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java (hunk 4 above)
Comment (suprajkolluri) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:156]: engine.getTemplate() and engine.init() methods are throwing 'Exception', hence they are being caught.

Comment (suprajkolluri) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/rest/ConceptCollectionRestController.java:156]: Hey Julia, in some rest controllers we are catching the Exception and throwing RestException. In some controllers we are not catching the Exception class but throwing it directly. Which is a better practice?

File: Quadriga/src/main/resources/locale/errormessages.properties (hunk 17 above)
Comment (suprajkolluri) [inline Quadriga/src/main/resources/locale/errormessages.properties:5]: This was written in QuadrigaRestExceptionHandler
if(errorcode==0)
errorcode=406;
To may be say that if there was no error code set, give this value.
This is not thrown from anywhere though.

**Pull request closed without merging**